### PR TITLE
Metric guide & model_eval checks refactor

### PR DESCRIPTION
**Pull request by noamzbr**

### CODE REVIEW

File: docs/source/checks/vision/model_evaluation/plot_class_performance.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 * `What is the purpose of the check? <#what-is-the-purpose-of-the-check>`__
 * `Classification <#classification-performance-report>`__
 
-  - `Generate data & model <#generate-data-and-model>`__
+  - `Generate data set <#generate-dataset>`__
```
Comment (noamzbr) [inline docs/source/checks/vision/model_evaluation/plot_class_performance.py:15]:
```suggestion
  - `Generate Dataset <#generate-dataset>`__
```

File: docs/source/checks/vision/model_evaluation/plot_class_performance.py [hunk 2]
```diff
@@ -35,29 +35,31 @@
...
 
 from deepchecks.vision.checks import ClassPerformance
 from deepchecks.vision.datasets.classification import mnist_torch as mnist
 
 #%%
 # Classification Performance Report
 # =================================
-# Generate data and model:
-# ------------------------
+# Generate dataset:
```
Comment (noamzbr) [inline docs/source/checks/vision/model_evaluation/plot_class_performance.py:59]:
```suggestion
# Generate Dataset:
```

File: docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py [hunk 3]
```diff
@@ -10,47 +10,46 @@
...
-* `Generate data an model <#generate-data-and-model>`__
+* `Generate Dataset <#generate-dataset>`__
 * `Run the check <#run-the-check>`__
 * `Define a condition <#define-a-condition>`__
 
 What Is the Purpose of the Check?
 =================================
 This check returns the results from a dict of metrics, in the format metric name: scorer, calculated for the
-given model dataset. The scorer can be an ignite.Metric or Sklearn scorer.
+given model dataset. The scorer should be either a sklearn scorer or a (see :doc:`` for further details).
```
Comment (Nadav-Barak) [inline docs/source/checks/vision/model_evaluation/plot_single_dataset_performance.py:20]:
```suggestion
given model dataset. The scorer should be either a sklearn scorer or a :doc:`custom scorer<link>'.
```

File: examples/examples_metrics_guide.py [hunk 4]
```diff
@@ -31,23 +22,56 @@
...
+from ignite.metrics import Metric
+from ignite.metrics.metric import reinit__is_reduced, sync_all_reduce
 from deepchecks.vision.checks import SingleDatasetPerformance
 
-precision = Precision(average=True)
-double_precision = 2 * precision
+# For simplicity, we will implement the accuracy metric, although it is already implemented in deepchecks and
+# can be passed as a string, and even otherwise we'd recommend using the sklearn API for custom classification metrics.
+
+class CustomAccuracy(Metric):
```
Comment (Nadav-Barak) [inline examples/examples_metrics_guide.py:42]: Damn thats complicated

File: docs/source/user-guide/general/metrics_guide.rst [hunk 5]
```diff
@@ -259,11 +261,25 @@ Custom Metrics
...
+The ``update`` method is called on each batch of data, and the ``compute`` method is called to compute the final metric.
+Note that contrary to the natively implemented ignite metrics, which expect the the labels and predictions to be
+pytorch tensors, your custom metrics should expect lists of numpy objets. For example, for classification the labels
+will be a list of integers, while the predictions will be lists of numpy array representing the softmax outputs. This is
+designed to enable deepcheck to function regardless of the underlying framework.
+
+Fir the same reason, for classification you shouldn't implement custom Ignite Metrics, but rather use the (simpler)
+Scikit-learn Scorer API which deepchecks supports natively also for computer vision tasks. Ignite metrics are
+recommended only if you haven't found a suitable metric in the list of deepchecks metrics above, or for new tasks that
+are not yet supported by deepchecks.
```
Comment (Nadav-Barak) [inline docs/source/user-guide/general/metrics_guide.rst:281]: maybe describe CustomClassificationScorer which is a useful for custom classification scorers

Comment (Nadav-Barak) [reply docs/source/user-guide/general/metrics_guide.rst:281]: alternative is to describe here you can just send a custom callable + edit stuff in code - but i guess we do not want to do that

**Review by Nadav-Barak — COMMENTED** [review]
Custom scorer is very very complicated :(
Also apply your own comments!

File: docs/source/checks/vision/model_evaluation/plot_class_performance.py (hunk 1 above)
Comment (noamzbr) [reply docs/source/checks/vision/model_evaluation/plot_class_performance.py:15]: deal with it

File: deepchecks/vision/metrics_utils/scorers.py [hunk 6]
```diff
@@ -140,9 +140,15 @@ def get_scorers_dict(
                     raise DeepchecksNotSupportedError(
                         f'Unsupported metric: {name} of type {type(metric).__name__} was given.')
                 scorers[name] = converted_met
+            elif isinstance(metric, t.Callable):
+                if task_type == TaskType.CLASSIFICATION:
+                    scorers[name] = CustomClassificationScorer(metric)
+                else:
+                    raise DeepchecksNotSupportedError('Custom scikit-learn scorers are only supported for'
```
Comment (Nadav-Barak) [inline deepchecks/vision/metrics_utils/scorers.py:147]: its not just scikit-learn scorers, all callables only work in classification

File: examples/examples_metrics_guide.py [hunk 7]
```diff
@@ -31,23 +22,51 @@
...
+
+    def __init__(self):
+        super().__init__()
+
+    def reset(self):
+        self._correct = 0
+        self._total = 0
+        super().reset()
+
+    def update(self, output):
```
Comment (Nadav-Barak) [inline examples/examples_metrics_guide.py:50]: add type declarations, output is a tuple of numpy here? can verify in checks

File: docs/source/user-guide/general/metrics_guide.rst [hunk 8]
```diff
@@ -259,12 +261,27 @@ Custom Metrics
...
+a function that accepts the parameters: (model, x, y_true), and returns a score with the convention that higher is
+better.
+
+For other computer vision tasks, the custom metrics should be a custom metric class. Deepchecks Custom Metric classes
+are inspired by `Ignite metrics <https://pytorch.org/ignite/metrics.html#how-to-create-a-custom-metric>`__ and implement
+the following methods: ``reset``, ``update`` and ``compute``.
+
+    * ``reset`` - resets the metric to its initial state, creates or resets any internal variables.
+    * ``update`` - updates the metric's internal state based on the provided. The method's signature
+        should be ``update(self, y_pred, y_true)``, where ``y_pred`` is the model's output and ``y_true`` is the ground
```
Comment (Nadav-Barak) [inline docs/source/user-guide/general/metrics_guide.rst:275]: this is not the case in your example

File: docs/source/user-guide/general/metrics_guide.rst [hunk 9]
```diff
@@ -259,12 +261,27 @@ Custom Metrics
...
+        truth, both given  as lists of numpy objets. For example, for object detection the label would be a list
+        where each element is a numpy array of bounding boxes annotations, and the prediction would be a list where each
+        element is a numpy array of bounding boxes predictions, both in the
+        :doc:`deepchecks format </user-guide/vision/supported_tasks_and_formats.rst>`.
+    * ``compute`` - returns the metric's value based on the internal state.
```
Comment (Nadav-Barak) [inline docs/source/user-guide/general/metrics_guide.rst:280]: should return either a single number of a list of numbers representing value per class.
maybe also mention they must be numpy or int (not tensors)

File: deepchecks/vision/metrics_utils/scorers.py (hunk 6 above)
Comment (noamzbr) [reply deepchecks/vision/metrics_utils/scorers.py:147]: we don't want them passing any callable that is not an sklearn scorer. Can't enforce that though.

File: deepchecks/vision/metrics_utils/scorers.py [hunk 10]
```diff
@@ -140,9 +140,15 @@ def get_scorers_dict(
                     raise DeepchecksNotSupportedError(
                         f'Unsupported metric: {name} of type {type(metric).__name__} was given.')
                 scorers[name] = converted_met
+            elif isinstance(metric, t.Callable):
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/vision/metrics_utils/scorers.py:143]: Added CustomMetric

File: docs/source/user-guide/general/metrics_guide.rst (hunk 8 above)
Comment (noamzbr) [reply docs/source/user-guide/general/metrics_guide.rst:275]: fixed

File: docs/source/user-guide/general/metrics_guide.rst [hunk 11]
```diff
@@ -259,12 +261,29 @@ Custom Metrics
...
+better.
+
+For other computer vision tasks, the custom metrics should be a custom metric class. Deepchecks Custom Metric classes
+are inspired by `Ignite metrics <https://pytorch.org/ignite/metrics.html#how-to-create-a-custom-metric>`__, must
+inherit from :class:`deepchecks.vision.metric_utils.custom_metric.CustomMetric` and implement the following methods:
+``reset``, ``update`` and ``compute``.
+
+    * ``reset`` - resets the metric to its initial state, creates or resets any internal variables.
+    * ``update`` - updates the metric's internal state based on the provided labels and predictions. The method's
+        signature should be ``update(self, output)``, where output is a tuple containing first ``y_pred`` which is the
```
Comment (Nadav-Barak) [inline docs/source/user-guide/general/metrics_guide.rst:276]: thats how ignite metrics work? otherwise I think update(self, y_true,y_pred) is better as convention

File: docs/source/user-guide/general/metrics_guide.rst [hunk 12]
```diff
@@ -259,12 +261,29 @@ Custom Metrics
...
+        model's output and second ``y_true`` is the ground truth, both given  as lists of numpy objets. For example, for
+        object detection the label would be a list where each element is a numpy array of bounding boxes annotations,
+        and the prediction would be a list where each element is a numpy array of bounding boxes predictions, both in
+        the :doc:`deepchecks format </user-guide/vision/supported_tasks_and_formats.rst>`.
+    * ``compute`` - returns the metric's value based on the internal state. Can be either a single number, or a numpy
+        array of metrics values per class.
```
Comment (Nadav-Barak) [inline docs/source/user-guide/general/metrics_guide.rst:282]: what do you mean by array of metrics values?

File: deepchecks/vision/metrics_utils/scorers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/vision/metrics_utils/scorers.py:143]: Yep thats reasonable

[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/vision/metrics_utils/scorers.py:143]: Would add additional validation for this if (validate that the callable is in the required format, maybe this can be done in DeepchecksScorer now that i think about it

File: docs/source/user-guide/general/metrics_guide.rst (hunk 11 above)
Comment (noamzbr) [reply docs/source/user-guide/general/metrics_guide.rst:276]: yep that's how ignite works

File: deepchecks/vision/metrics_utils/scorers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/vision/metrics_utils/scorers.py:143]: As this is a problem with tabular as well, would do in PR unrelated to #2211

File: docs/source/user-guide/general/metrics_guide.rst (hunk 12 above)
Comment (noamzbr) [reply docs/source/user-guide/general/metrics_guide.rst]: Tried an alternative phrasing, take a look

File: deepchecks/vision/metrics_utils/scorers.py (hunk 6 above)
Comment (Nadav-Barak) [reply deepchecks/vision/metrics_utils/scorers.py:147]: IMO you want all callable in sklearn format (not necessarily just stuff imported fro sklearn)
Can maybe just check that the callable expects the relevant arguments? not that important

File: deepchecks/vision/metrics_utils/scorers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/vision/metrics_utils/scorers.py:143]: sure

File: docs/source/user-guide/general/metrics_guide.rst (hunk 11 above)
Comment (Nadav-Barak) [reply docs/source/user-guide/general/metrics_guide.rst:276]: :(

File: docs/source/user-guide/general/metrics_guide.rst [hunk 13]
```diff
@@ -259,12 +261,29 @@ Custom Metrics
...
-`sklearn scorer <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html>`__ API.
-Multiclass classification scorers should assume that the labels are given in a
+For tabular metrics and vision classification tasks the custom metrics function should follow the
+`sklearn scorer <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html>`__ API, which is
+a function that accepts the parameters: (model, x, y_true), and returns a score with the convention that higher is
+better.
+
+For other computer vision tasks, the custom metrics should be a custom metric class. Deepchecks Custom Metric classes
+are inspired by `Ignite metrics <https://pytorch.org/ignite/metrics.html#how-to-create-a-custom-metric>`__, must
+inherit from :class:`deepchecks.vision.metric_utils.custom_metric.CustomMetric` and implement the following methods:
```
Comment (Nadav-Barak) [inline docs/source/user-guide/general/metrics_guide.rst:271]: Dont really like the phresing here, specifically the first sentence

File: docs/source/user-guide/general/metrics_guide.rst [hunk 14]
```diff
@@ -259,12 +261,29 @@ Custom Metrics
...
+``reset``, ``update`` and ``compute``.
+
+    * ``reset`` - resets the metric to its initial state, creates or resets any internal variables.
```
Comment (Nadav-Barak) [inline docs/source/user-guide/general/metrics_guide.rst:274]: creation of variable happen at init

File: docs/source/user-guide/general/metrics_guide.rst [hunk 15]
```diff
@@ -259,12 +261,29 @@ Custom Metrics
...
+    * ``update`` - updates the metric's internal state based on the provided labels and predictions. The method's
+        signature should be ``update(self, output)``, where output is a tuple containing first ``y_pred`` which is the
+        model's output and second ``y_true`` is the ground truth, both given  as lists of numpy objets. For example, for
```
Comment (Nadav-Barak) [inline docs/source/user-guide/general/metrics_guide.rst:277]: Is it the model output or batch.numpy_predictions?

File: examples/examples_metrics_guide.py [hunk 16]
```diff
@@ -31,23 +22,53 @@
...
+
+class CustomAccuracy(CustomMetric):
+
+    def __init__(self):
+        super().__init__()
+
+    def reset(self):
+        self._correct = 0
+        self._total = 0
+        super().reset()
```
Comment (Nadav-Barak) [inline examples/examples_metrics_guide.py:50]: why the super in the methods? (not saying i'm against it)

**Review by Nadav-Barak — APPROVED** [review]
